homeassistant/components/nissan_leaf/__init__.py

```diff
@@ -331,151 +310,146 @@ def _extract_start_date(battery_info):
         except KeyError:
             return None
 
-    async def async_get_battery(self):
+    def get_battery(self):
         """Request battery update from Nissan servers."""
         from pycarwings2 import CarwingsError
         try:
             # First, check nissan servers for the latest data
-            start_server_info = await self.hass.async_add_job(
-                self.leaf.get_latest_battery_status
-            )
+            start_server_info = self.leaf.get_latest_battery_status()
 
             # Store the date from the nissan servers
             start_date = self._extract_start_date(start_server_info)
             if start_date is None:
                 _LOGGER.info("No start date from servers. Aborting")
                 return None
 
-            _LOGGER.info("Start server date=%s", start_date)
+            _LOGGER.debug("Start server date=%s", start_date)
 
             # Request battery update from the car
-            _LOGGER.info("Requesting battery update, %s", self.leaf.vin)
-            request = await self.hass.async_add_job(self.leaf.request_update)
+            _LOGGER.debug("Requesting battery update, %s", self.leaf.vin)
+            request = self.leaf.request_update()
             if not request:
                 _LOGGER.error("Battery update request failed")
                 return None
 
             for attempt in range(MAX_RESPONSE_ATTEMPTS):
-                _LOGGER.info("Waiting %s seconds for battery update (%s) (%s)",
-                             PYCARWINGS2_SLEEP, self.leaf.vin, attempt)
-                await asyncio.sleep(PYCARWINGS2_SLEEP)
+                _LOGGER.debug(
+                    "Waiting %s seconds for battery update (%s) (%s)",
+                    PYCARWINGS2_SLEEP, self.leaf.vin, attempt)
+                # Not sure of the best way to sleep in sync mode,
+                # used to be asyncio.sleep
+                time.sleep(PYCARWINGS2_SLEEP)
```

Comment:
Sync sleep is not ok, so here we have to use asyncio to sleep. So all the methods that need to sleep need to be coroutines.

Reply:
Attempted fix in commit https://github.com/home-assistant/home-assistant/pull/21145/commits/df448adf57edbed553969b3a2bd76da707dede21, this forces a lot of code to async.  I still have concerns over use of `dispatcher_connect` and `dispatcher_send`.

Reply:
Resolved issues with `dispatcher_connect` and `dispatcher_send` in other commits.